module/zfs/dbuf.c

```diff
@@ -2344,20 +2344,19 @@ dmu_buf_will_dirty_impl(dmu_buf_t *db_fake, int flags, dmu_tx_t *tx)
 	 */
 	mutex_enter(&db->db_mtx);
 
-	dbuf_dirty_record_t *dr;
-	for (dr = db->db_last_dirty;
-	    dr != NULL && dr->dr_txg >= tx->tx_txg; dr = dr->dr_next) {
-		/*
-		 * It's possible that it is already dirty but not cached,
-		 * because there are some calls to dbuf_dirty() that don't
-		 * go through dmu_buf_will_dirty().
-		 */
-		if (dr->dr_txg == tx->tx_txg && db->db_state == DB_CACHED) {
-			/* This dbuf is already dirty and cached. */
-			dbuf_redirty(dr);
-			mutex_exit(&db->db_mtx);
-			return;
-		}
+	dbuf_dirty_record_t *dr = NULL;
+	if (db->db_state == DB_CACHED)
+		dr = dbuf_find_dirty_eq(db, tx->tx_txg);
+	/*
+	 * It's possible that it is already dirty but not cached,
+	 * because there are some calls to dbuf_dirty() that don't
+	 * go through dmu_buf_will_dirty().
+	 */
+	if (dr) {
+		/* This dbuf is already dirty and cached. */
+		dbuf_redirty(dr);
+		mutex_exit(&db->db_mtx);
```

Comment:
I think this would be more clear as:
```
if (state==CACHED) {
    dbuf_dirty_record_t *dr = dbuf_find_dirty_eq(...)
    if (dr != NULL) {
      ...
      return;
    }
}
```